File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/websocket/SocketMessage.java
package io.casey.musikcube.remote.websocket;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Locale;
import java.util.concurrent.atomic.AtomicInteger;

public class SocketMessage {
    private static final String TAG = SocketMessage.class.getCanonicalName();

    public enum Type {
        Request("request"),
        Response("response"),
        Broadcast("broadcast");

        private String rawType;

        Type(String rawType) {
            this.rawType = rawType;
        }

        public String getRawType() {
            return rawType;
        }

        public static Type fromString(final String str) {
            if (Request.rawType.equals(str)) {
                return Request;
            }
            else if (Response.rawType.equals(str)) {
                return Response;
            }
            else if (Broadcast.rawType.equals(str)) {
                return Broadcast;
            }

            throw new IllegalArgumentException("str");
        }
    }

    private String name;
    private String id;
    private Type type;
    private JSONObject options;

    public static SocketMessage create(String string) {
        try {
            JSONObject object = new JSONObject(string);
            final String name = object.getString("name");
            final Type type = Type.fromString(object.getString("type"));
            final String id = object.getString("id");
            final JSONObject options = object.optJSONObject("options");
            return new SocketMessage(name, id, type, options);
        }
        catch (Exception ex) {
            Log.e(TAG, ex.toString());
            return null;
        }
    }

    private SocketMessage(String name, String id, Type type, JSONObject options) {
        if (name == null || name.length() == 0 || id == null || id.length() == 0 || type == null) {
            throw new IllegalArgumentException();
        }

        this.name = name;
        this.id = id;
        this.type = type;
        this.options = (options == null) ? new JSONObject() : options;
    }

    public String getName() {
        return name;
    }

    public String getId() {
        return id;
    }

    public Type getType() {
        return type;
    }

    public <T> T getOption(final String key) {
        if (options.has(key)) {
            try {
                return (T) options.get(key);
            }
            catch (JSONException ex) {
                /* swallow */
            }
        }
        return null;
    }

    public String getStringOption(final String key) {
        return getStringOption(key, "");
    }

    public String getStringOption(final String key, final String defaultValue) {
        if (options.has(key)) {
            try {
                return options.getString(key);
            }
            catch (JSONException ex) {
                /* swallow */
            }
        }
        return defaultValue;
    }

    public int getIntOption(final String key) {
        return getIntOption(key, 0);
    }

    public int getIntOption(final String key, final int defaultValue) {
        if (options.has(key)) {
            try {
                return options.getInt(key);
            }
            catch (JSONException ex) {
                /* swallow */
            }
        }
        return defaultValue;
    }


    public double getDoubleOption(final String key) {
        return getDoubleOption(key, 0.0);
    }

    public double getDoubleOption(final String key, final double defaultValue) {
        if (options.has(key)) {
            try {
                return options.getDouble(key);
            }
            catch (JSONException ex) {
                /* swallow */
            }
        }
        return defaultValue;
    }

    public boolean getBooleanOption(final String key) {
        return getBooleanOption(key, false);
    }

    public boolean getBooleanOption(final String key, final boolean defaultValue) {
        if (options.has(key)) {
            try {
                return options.getBoolean(key);
            }
            catch (JSONException ex) {
                /* swallow */
            }
        }
        return defaultValue;
    }

    public JSONObject getJsonObjectOption(final String key) {
        return getJsonObjectOption(key, null);
    }

    public JSONObject getJsonObjectOption(final String key, final JSONObject defaultValue) {
        if (options.has(key)) {
            try {
                return options.getJSONObject(key);
            }
            catch (JSONException ex) {
                /* swallow */
            }
        }
        return defaultValue;
    }

    public JSONArray getJsonArrayOption(final String key) {
        return getJsonArrayOption(key, null);
    }

    public JSONArray getJsonArrayOption(final String key, final JSONArray defaultValue) {
        if (options.has(key)) {
            try {
                return options.getJSONArray(key);
            }
            catch (JSONException ex) {
                /* swallow */
            }
        }
        return defaultValue;
    }

    @Override
    public String toString() {
        try {
            final JSONObject json = new JSONObject();
            json.put("name", name);
            json.put("id", id);
            json.put("type", type.getRawType());
            json.put("options", options);
            return json.toString();
        }
        catch (JSONException ex) {
            throw new RuntimeException("unable to generate output JSON!");
        }
    }

    public Builder buildUpon() {
        try {
            final Builder builder = new Builder();
            builder.name = name;
            builder.type = type;
            builder.id = id;
            builder.options = new JSONObject(options.toString());
            return builder;
        }
        catch (JSONException ex) {
            throw new RuntimeException(ex);
        }
    }

    public static class Builder {
        private static AtomicInteger nextId = new AtomicInteger();

        private String name;
        private Type type;
        private String id;
        private JSONObject options = new JSONObject();

        private Builder() {

        }

        private static String newId() {
            return String.format(Locale.ENGLISH, "musikcube-android-client-%d", nextId.incrementAndGet());
        }

        public static Builder broadcast(String name) {
            final Builder builder = new Builder();
            builder.name = name;
            builder.id = newId();
            builder.type = Type.Response;
            return builder;
        }

        public static Builder respondTo(final SocketMessage message) {
            final Builder builder = new Builder();
            builder.name = message.getName();
            builder.id = message.getId();
            builder.type = Type.Response;
            return builder;
        }

        public static Builder request(final String name) {
            final Builder builder = new Builder();
            builder.name = name;
            builder.id = newId();
            builder.type = Type.Request;
            return builder;
        }

        public static Builder request(final Messages.Request name) {
            final Builder builder = new Builder();
            builder.name = name.toString();
            builder.id = newId();
            builder.type = Type.Request;
            return builder;
        }

        public Builder addOption(final String key, final Object value) {
            try {
                options.put(key, value);
            }
            catch (JSONException ex) {
                throw new RuntimeException("addOption failed??");
            }
            return this;
        }

        public Builder removeOption(final String key) {
            options.remove(key);
            return this;
        }

        public SocketMessage build() {
            return new SocketMessage(name, id, type, options);
        }
    }
}

File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/service/websocket/SocketMessage.kt
package io.casey.musikcube.remote.service.websocket

import android.util.Log
import io.casey.musikcube.remote.Application
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class SocketMessage private constructor(val name: String, val id: String, val type: SocketMessage.Type, options: JSONObject?) {
    enum class Type constructor(val rawType: String) {
        Request("request"),
        Response("response"),
        Broadcast("broadcast");

        companion object {
            fun fromString(str: String): Type =
                when (str) {
                    Request.rawType -> Request
                    Response.rawType -> Response
                    Broadcast.rawType -> Broadcast
                    else -> throw IllegalArgumentException("str")
                }
        }
    }

    private val options: JSONObject

    init {
        if (name.isEmpty() || id.isEmpty()) {
            throw IllegalArgumentException()
        }
        this.options = options ?: JSONObject()
    }

    @Suppress("UNCHECKED_CAST", "unused")
    fun <T> getOption(key: String): T? {
        if (options.has(key)) {
            try {
                return options.get(key) as T
            }
            catch (ex: JSONException) {
                /* swallow */
            }
        }
        return null
    }

    @JvmOverloads fun getStringOption(key: String, defaultValue: String = ""): String {
        if (options.has(key)) {
            try {
                return options.getString(key)
            }
            catch (ex: JSONException) {
                /* swallow */
            }
        }
        return defaultValue
    }

    @JvmOverloads fun getIntOption(key: String, defaultValue: Int = 0): Int {
        if (options.has(key)) {
            try {
                return options.getInt(key)
            }
            catch (ex: JSONException) {
                /* swallow */
            }
        }
        return defaultValue
    }

    @JvmOverloads fun getLongOption(key: String, defaultValue: Long = 0): Long {
        if (options.has(key)) {
            try {
                return options.getLong(key)
            }
            catch (ex: JSONException) {
                /* swallow */
            }
        }
        return defaultValue
    }

    @JvmOverloads fun getDoubleOption(key: String, defaultValue: Double = 0.0): Double {
        if (options.has(key)) {
            try {
                return options.getDouble(key)
            }
            catch (ex: JSONException) {
                /* swallow */
            }
        }
        return defaultValue
    }

    @JvmOverloads fun getBooleanOption(key: String, defaultValue: Boolean = false): Boolean {
        if (options.has(key)) {
            try {
                return options.getBoolean(key)
            }
            catch (ex: JSONException) {
                /* swallow */
            }
        }
        return defaultValue
    }

    fun getJsonObject(): JSONObject {
        return JSONObject(options.toString())
    }

    @JvmOverloads fun getJsonObjectOption(key: String, defaultValue: JSONObject? = null): JSONObject? {
        if (options.has(key)) {
            try {
                return options.getJSONObject(key)
            }
            catch (ex: JSONException) {
                /* swallow */
            }
        }
        return defaultValue
    }

    @JvmOverloads fun getJsonArrayOption(key: String, defaultValue: JSONArray? = null): JSONArray? {
        if (options.has(key)) {
            try {
                return options.getJSONArray(key)
            }
            catch (ex: JSONException) {
                /* swallow */
            }
        }
        return defaultValue
    }

    override fun toString(): String {
        try {
            val json = JSONObject()
            json.put("name", name)
            json.put("id", id)
            json.put("device_id", Application.deviceId)
            json.put("type", type.rawType)
            json.put("options", options)
            return json.toString()
        }
        catch (ex: JSONException) {
            throw RuntimeException("unable to generate output JSON!")
        }
    }

    @Suppress("unused")
    fun buildUpon(): Builder {
        try {
            val builder = Builder()
            builder.internalName = name
            builder.internalType = type
            builder.internalId = id
            builder.internalOptions = JSONObject(options.toString())
            return builder
        }
        catch (ex: JSONException) {
            throw RuntimeException(ex)
        }
    }

    class Builder internal constructor() {
        internal var internalName: String? = null
        internal var internalType: Type? = null
        internal var internalId: String? = null
        internal var internalOptions = JSONObject()

        fun withOptions(options: JSONObject?): Builder {
            internalOptions = options ?: JSONObject()
            return this
        }

        fun addOption(key: String, value: Any?): Builder {
            try {
                internalOptions.put(key, value)
            }
            catch (ex: JSONException) {
                throw RuntimeException("addOption failed??")
            }
            return this
        }

        @Suppress("unused")
        fun removeOption(key: String): Builder {
            internalOptions.remove(key)
            return this
        }

        fun build(): SocketMessage {
            return SocketMessage(internalName!!, internalId!!, internalType!!, internalOptions)
        }

        companion object {
            private val nextId = AtomicInteger()

            private fun newId(): String {
                return String.format(Locale.ENGLISH, "musikcube-android-client-%d", nextId.incrementAndGet())
            }

            @Suppress("unused")
            fun broadcast(name: String): Builder {
                val builder = Builder()
                builder.internalName = name
                builder.internalId = newId()
                builder.internalType = Type.Response
                return builder
            }

            fun respondTo(message: SocketMessage): Builder {
                val builder = Builder()
                builder.internalName = message.name
                builder.internalId = message.id
                builder.internalType = Type.Response
                return builder
            }

            fun request(name: String): Builder {
                val builder = Builder()
                builder.internalName = name
                builder.internalId = newId()
                builder.internalType = Type.Request
                return builder
            }

            fun request(name: Messages.Request): Builder {
                val builder = Builder()
                builder.internalName = name.toString()
                builder.internalId = newId()
                builder.internalType = Type.Request
                return builder
            }
        }
    }

    companion object {
        private val TAG = SocketMessage::class.java.canonicalName

        fun create(string: String): SocketMessage? =
            try {
                val `object` = JSONObject(string)
                val name = `object`.getString("name")
                val type = Type.fromString(`object`.getString("type"))
                val id = `object`.getString("id")
                val options = `object`.optJSONObject("options")
                SocketMessage(name, id, type, options)
            }
            catch (ex: Exception) {
                Log.e(TAG, ex.toString())
                null
            }
    }
}
